core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -106,7 +107,66 @@
         /**
          * Unassigned as a result of a failed primary while the replica was initializing.
          */
-        PRIMARY_FAILED;
+        PRIMARY_FAILED
+    }
+
+    /**
+     * Captures the status of an unsuccessful allocation attempt for the shard,
+     * causing it to remain in the unassigned state.
+     *
+     * Note, ordering of the enum is important, make sure to add new values
+     * at the end and handle version serialization properly.
+     */
+    public enum AllocationAttempt implements Writeable {
+        /**
+         * The shard was denied allocation to a node because the allocation deciders all returned a NO decision
+         */
+        DENIED_BY_ALLOCATION_DECIDERS,
+        /**
+         * The shard was denied allocation to a node because there were no valid shard copies found for it;
+         * this can happen on node restart with gateway allocation
+         */
+        NO_VALID_SHARD_COPY,
+        /**
+         * The allocation attempt was throttled on the shard by the allocation deciders
+         */
+        THROTTLED_BY_ALLOCATION_DECIDERS,
+        /**
+         * Waiting on getting shard data from all nodes before making a decision about where to allocate the shard
+         */
+        FETCHING_SHARD_DATA,
+        /**
+         * Allocation decision has been delayed
+         */
+        DELAYED_ALLOCATION,
+        /**
+         * No allocation attempt has been made yet
+         */
+        NO_ATTEMPT;
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeByte((byte) ordinal());
+        }
+
+        public static AllocationAttempt readFrom(StreamInput in) throws IOException {
+            return values()[(int) in.readByte()];
+        }
+
+        public static AllocationAttempt fromDecision(Decision decision) {
+            if (decision == null) {
+                return DENIED_BY_ALLOCATION_DECIDERS;
```

Comment:
I prefer if null is handled outside of this method (there is just one usage atm.)